File: architecture-examples/monolithic.py
"""
Monolithic Architecture - Python

Single, unified application where all components are tightly coupled
and deployed together as one unit.
"""


# Monolithic Application Structure
class MonolithicApp:
    def __init__(self):
        self.user_service = UserService()
        self.order_service = OrderService()
        self.payment_service = PaymentService()
        self.database = Database()
    
    # All services share the same database connection
    def initialize(self):
        self.database.connect()
        print("Monolithic app initialized")
    
    def handle_request(self, endpoint, data):
        if endpoint == '/users':
            return self.user_service.handle(data)
        elif endpoint == '/orders':
            return self.order_service.handle(data)
        elif endpoint == '/payments':
            return self.payment_service.handle(data)
        else:
            return {'error': 'Not found'}


class UserService:
    def __init__(self):
        self.db = None  # Shared database
    
    def set_database(self, db):
        self.db = db
    
    def handle(self, data):
        if data.get('action') == 'create':
            return self.create_user(data.get('user'))
        return {'error': 'Invalid action'}
    
    def create_user(self, user):
        # Direct database access
        import time
        return {'id': int(time.time()), **user, 'created': time.time()}


class OrderService:
    def __init__(self):
        self.db = None
    
    def set_database(self, db):
        self.db = db
    
    def handle(self, data):
        if data.get('action') == 'create':
            return self.create_order(data.get('order'))
        return {'error': 'Invalid action'}
    
    def create_order(self, order):
        import time
        return {'id': int(time.time()), **order, 'status': 'pending'}


class PaymentService:
    def handle(self, data):
        if data.get('action') == 'process':
            return self.process_payment(data.get('payment'))
        return {'error': 'Invalid action'}
    
    def process_payment(self, payment):
        import time
        return {'id': int(time.time()), **payment, 'status': 'completed'}


class Database:
    def connect(self):
        print("Database connected (shared connection)")


# Example usage
if __name__ == "__main__":
    print("=== Monolithic Architecture ===\n")
    
    app = MonolithicApp()
    app.initialize()
    
    print("\nHandling requests:")
    print(app.handle_request('/users', {'action': 'create', 'user': {'name': 'John'}}))
    print(app.handle_request('/orders', {'action': 'create', 'order': {'item': 'Book'}}))
    print(app.handle_request('/payments', {'action': 'process', 'payment': {'amount': 100}}))

File: architecture-examples/monolithic.js
/**
 * Monolithic Architecture - JavaScript
 * 
 * Single, unified application where all components are tightly coupled
 * and deployed together as one unit.
 */

// Monolithic Application Structure
class MonolithicApp {
    constructor() {
        this.userService = new UserService();
        this.orderService = new OrderService();
        this.paymentService = new PaymentService();
        this.database = new Database();
    }
    
    // All services share the same database connection
    initialize() {
        this.database.connect();
        console.log('Monolithic app initialized');
    }
    
    handleRequest(endpoint, data) {
        switch (endpoint) {
            case '/users':
                return this.userService.handle(data);
            case '/orders':
                return this.orderService.handle(data);
            case '/payments':
                return this.paymentService.handle(data);
            default:
                return { error: 'Not found' };
        }
    }
}

class UserService {
    constructor() {
        this.db = null; // Shared database
    }
    
    setDatabase(db) {
        this.db = db;
    }
    
    handle(data) {
        if (data.action === 'create') {
            return this.createUser(data.user);
        }
        return { error: 'Invalid action' };
    }
    
    createUser(user) {
        // Direct database access
        return { id: Date.now(), ...user, created: new Date() };
    }
}

class OrderService {
    constructor() {
        this.db = null;
    }
    
    setDatabase(db) {
        this.db = db;
    }
    
    handle(data) {
        if (data.action === 'create') {
            return this.createOrder(data.order);
        }
        return { error: 'Invalid action' };
    }
    
    createOrder(order) {
        return { id: Date.now(), ...order, status: 'pending' };
    }
}

class PaymentService {
    handle(data) {
        if (data.action === 'process') {
            return this.processPayment(data.payment);
        }
        return { error: 'Invalid action' };
    }
    
    processPayment(payment) {
        return { id: Date.now(), ...payment, status: 'completed' };
    }
}

class Database {
    connect() {
        console.log('Database connected (shared connection)');
    }
}

// Example usage
console.log('=== Monolithic Architecture ===\n');

const app = new MonolithicApp();
app.initialize();

console.log('\nHandling requests:');
console.log(app.handleRequest('/users', { action: 'create', user: { name: 'John' } }));
console.log(app.handleRequest('/orders', { action: 'create', order: { item: 'Book' } }));
console.log(app.handleRequest('/payments', { action: 'process', payment: { amount: 100 } }));

module.exports = { MonolithicApp, UserService, OrderService, PaymentService };

